src/vs/workbench/contrib/interactiveSession/browser/actions/interactiveSessionTitleActions.ts

```diff
@@ -90,4 +97,124 @@ export function registerInteractiveSessionTitleActions() {
 		}
 	});
+
+	registerAction2(class InsertToNotebookAction extends Action2 {
+		constructor() {
+			super({
+				id: 'workbench.action.interactiveSession.insertIntoNotebook',
+				title: {
+					value: localize('interactive.insertIntoNotebook.label', "Insert into Notebook"),
+					original: 'Insert into Notebook'
+				},
+				f1: false,
+				category: INTERACTIVE_SESSION_CATEGORY,
+				icon: Codicon.insert,
+				menu: {
+					id: MenuId.InteractiveSessionTitle,
+					group: 'navigation',
+					isHiddenByDefault: true,
+					when: NOTEBOOK_IS_ACTIVE_EDITOR
+				}
+			});
+		}
+
+		async run(accessor: ServicesAccessor, ...args: any[]) {
+			const item = args[0];
+			if (!isResponseVM(item)) {
+				return;
+			}
+
+			const editorService = accessor.get(IEditorService);
+
+			if (editorService.activeEditorPane?.getId() === NOTEBOOK_EDITOR_ID) {
+				const notebookEditor = editorService.activeEditorPane.getControl() as INotebookEditor;
+
+				if (!notebookEditor.hasModel()) {
+					return;
+				}
+
+				if (notebookEditor.isReadOnly) {
+					return;
+				}
+
+				const value = item.response.value;
+				const splitContents = splitMarkdownAndCodeBlocks(value);
+
+				const focusRange = notebookEditor.getFocus();
+				const index = Math.max(focusRange.end, 0);
+				const bulkEditService = accessor.get(IBulkEditService);
+
+				await bulkEditService.apply(
+					[
+						new ResourceNotebookCellEdit(notebookEditor.textModel.uri,
+							{
+								editType: CellEditType.Replace,
+								index: index,
+								count: 0,
+								cells: splitContents.map(content => {
+									const kind = content.type === 'markdown' ? CellKind.Markup : CellKind.Code;
+									const language = content.type === 'markdown' ? 'markdown' : content.language;
+									const mime = content.type === 'markdown' ? 'text/markdown' : `text/x-${content.language}`;
+									return {
+										cellKind: kind,
+										language,
+										mime,
+										source: content.content,
+										outputs: [],
+										metadata: {}
+									};
+								})
+							}
+						)
+					],
+					{ quotableLabel: 'Insert into Notebook' }
+				);
+			}
+		}
+	});
+}
+
+interface MarkdownContent {
+	type: 'markdown';
+	content: string;
+}
+
+interface CodeContent {
+	type: 'code';
+	language: string;
+	content: string;
+}
+
+type Content = MarkdownContent | CodeContent;
+
+function splitMarkdownAndCodeBlocks(markdown: string): Content[] {
+	const lexer = new marked.Lexer();
+	const parser = new marked.Parser();
+	const tokens = lexer.lex(markdown);
+
+	const splitContent: Content[] = [];
+
+	let markdownPart = '';
+	tokens.forEach((token) => {
+		if (token.type === 'code') {
+			if (markdownPart.trim()) {
+				splitContent.push({ type: 'markdown', content: markdownPart });
+				markdownPart = '';
+			}
+			splitContent.push({
+				type: 'code',
+				language: token.lang || '',
+				content: token.text,
+			});
+		} else {
+			const text = parser.parse([token]);
```

Comment:
Don't you want the markdown source, which would be the raw text off the token?

Reply:
good catch, I should review carefully what copilot wrote.